### PR TITLE
[#157123071] Generate api typescript models

### PR DESCRIPTION
**Pull request by lussoluca**
Auto-generate models based on api.yaml

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,7 +16,8 @@
     "prettify": "prettier --write './**/*.ts'",
     "test": "jest -i",
     "start": "node src/server.js",
-    "generate-api-client": "swagger-cli bundle https://raw.githubusercontent.com/teamdigitale/digital-citizenship-functions/v0.38.0/api/public_api_v1.yaml | autorest --typescript --input-file=/dev/stdin --output-folder=src/api --add-credentials"
+    "generate-api-client": "swagger-cli bundle https://raw.githubusercontent.com/teamdigitale/digital-citizenship-functions/v0.38.0/api/public_api_v1.yaml | autorest --typescript --input-file=/dev/stdin --output-folder=src/api --add-credentials",
```
Comment (cloudify) [inline package.json]: btw for consistency in the other projects we use the `generate:` prefix

File: docker-compose.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ services:
     expose:
       - "80"
     ports:
-      - "81:80"
+      - "80:80"
```
Comment (cloudify) [inline docker-compose.yml]: what does this change have to do with "Generate api typescript models" ? :)

File: src/controllers/__tests__/profileController.test.ts [hunk 3]
```diff
@@ -20,7 +19,7 @@ const aFiscalNumber = "GRBGPP87L04L741X" as FiscalCode;
 const anEmailAddress = "garibaldi@example.com" as EmailAddress;
 const anIsInboxEnabled = true as IsInboxEnabled;
 const anIsWebookEnabled = true as IsWebhookEnabled;
-const aPreferredLanguage = "it_IT" as PreferredLanguage;
+const aPreferredLanguages: ReadonlyArray<any> = ["it_IT"];
```
Comment (cloudify) [inline src/controllers/__tests__/profileController.test.ts]: you should use the [PreferredLanguages](https://github.com/teamdigitale/digital-citizenship-functions/blob/master/api/definitions.yaml#L303) type here

File: src/controllers/notificationController.ts [hunk 4]
```diff
@@ -53,7 +55,7 @@ export default class NotificationController {
       return left(new Error("Unable to parse the installation ID"));
     }
 
-    const errorOrDevice = Device.decode(req.body);
+    const errorOrDevice = Installation.decode(req.body);
```
Comment (cloudify) [inline src/controllers/notificationController.ts]: `errorOrInstallation`

File: src/services/messagesService.ts [hunk 5]
```diff
@@ -6,17 +6,18 @@ import { isLeft } from "fp-ts/lib/Either";
 import { NonNegativeNumber } from "italia-ts-commons/lib/numbers";
 import { ReadableReporter } from "italia-ts-commons/lib/reporters";
 import * as winston from "winston";
+import { Messages } from "../types/api/Messages";
+import { MessageWithContent } from "../types/api/MessageWithContent";
 import { ProblemJson } from "../types/api/ProblemJson";
+import { ServicePublic as proxyServicePublic } from "../types/api/ServicePublic";
```
Comment (cloudify) [inline src/services/messagesService.ts]: type names should be capitalized

File: src/services/messagesService.ts [hunk 6]
```diff
@@ -6,17 +6,18 @@ import { isLeft } from "fp-ts/lib/Either";
...
 import { GetMessagesByUserOKResponse } from "../types/api_client/getMessagesByUserOKResponse";
 import { MessageResponseWithContent } from "../types/api_client/messageResponseWithContent";
-import { ServicePublic } from "../types/api_client/servicePublic";
+import { ServicePublic as apiServicePublic } from "../types/api_client/servicePublic";
```
Comment (cloudify) [inline src/services/messagesService.ts]: type names should be capitalized

File: src/services/notificationService.ts [hunk 7]
```diff
@@ -71,7 +71,7 @@ export default class NotificationService {
   public createOrUpdateInstallation(
     fiscalCode: FiscalCode,
     installationID: InstallationID,
-    device: Device
+    device: Installation
```
Comment (cloudify) [inline src/services/notificationService.ts]: `installation: Installation`

File: src/types/__tests__/profile.test.ts [hunk 8]
```diff
@@ -22,7 +21,7 @@ import { User } from "../user";
 const aFiscalNumber = "GRBGPP87L04L741X" as FiscalCode;
 const anEmailAddress = "garibaldi@example.com" as EmailAddress;
 const aNonNegativeNumber = 1 as NonNegativeNumber;
-const aPreferredLanguage = "it_IT" as PreferredLanguage;
+const aPreferredLanguages: ReadonlyArray<any> = ["it_IT"];
```
Comment (cloudify) [inline src/types/__tests__/profile.test.ts]: use `PreferredLanguages` type

**Comment by cloudify** [comment]
there are some types in `src/types/api_client` and some others in `src/types/api`, shouldn't all them be autogenerated and not committed in the repo? - cc @gunzip

File: docker-compose.yml (hunk 2 above)
Comment (lussoluca) [reply docker-compose.yml]: nothing :-(

**Comment by digitalcitizenship** [comment]
## Affected stories

  * ⚙️ [#157123071](https://www.pivotaltracker.com/story/show/157123071): Generare i modelli Typescript delle API esposte all'APP in automatico


New dependencies added: <a href='https://www.npmjs.com/package/italia-utils'><code>italia-utils</code></a>.


<h2><a href="https://github.com/teamdigitale/italia-utils#readme">italia-utils</a></h2>
<p>Author: https://teamdigitale.governo.it</p>
<p>Description: Tools and utilities for the Digital Citizenship project</p>
<p>Homepage: <a href="https://github.com/teamdigitale/italia-utils#readme">https://github.com/teamdigitale/italia-utils#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>24 days ago</td></tr><tr><td>Last Updated</td><td>15 minutes ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>2</td></tr><tr><td>Releases</td><td>16</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/fs-extra'><code>fs-extra</code></a>, <a href='https://www.npmjs.com/package/nunjucks'><code>nunjucks</code></a>, <a href='https://www.npmjs.com/package/prettier'><code>prettier</code></a>, <a href='https://www.npmjs.com/package/swagger-parser'><code>swagger-parser</code></a> and <a href='https://www.npmjs.com/package/swagger-schema-official'><code>swagger-schema-official</code></a></td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# Utilities and tools for the Digital Citizenship initiative

This package provide some tools that are used across the projects of the
Digital Citizenship initiative.

To add the tools to a project:

```
$ yarn add -D italia-utils
```

## gen-api-models

This tool generates TypeScript definitions of OpenAPI specs.

In simple terms it converts an OpenAPI spec like [this one](https://github.com/teamdigitale/digital-citizenship-functions/blob/f04666c8b7f2d4bebde19676b49b19119b03ef17/api/public_api_v1.yaml) into:

* A TypeScript [representation of the specs](https://github.com/teamdigitale/digital-citizenship-functions/blob/6798225bd725a42753b16375ce18a954a268f9b6/lib/api/public_api_v1.ts).
* An [io-ts](https://github.com/gcanti/io-ts) type definitions for [each API definition](https://github.com/teamdigitale/digital-citizenship-functions/tree/6798225bd725a42753b16375ce18a954a268f9b6/lib/api/definitions) that provides compile time types and runtime validation.

Note: the generated models requires the runtime dependency `italia-ts-commons`.

### Usage

```
$ gen-api-models --help
Options:
  --version       Show version number                                  [boolean]
  --api-spec      Path to input OpenAPI spec file            [string] [required]
  --out-dir       Output directory to store generated definition files
                                                             [string] [required]
  --ts-spec-file  If defined, converts the OpenAPI specs to TypeScript source
                  and writes it to this file                            [string]
  --help          Show help                                            [boolean]
```

Example:

```
$ gen-api-models --api-spec ./api/public_api_v1.yaml --out-dir ./lib/api/definitions --ts-spec-file ./lib/api/public_api_v1.ts
Writing TS Specs to lib/api/public_api_v1.ts
ProblemJson -> lib/api/definitions/ProblemJson.ts
NotificationChannel -> lib/api/definitions/NotificationChannel.ts
NotificationChannelStatusValue -> lib/api/definitions/NotificationChannelStatusValue.ts
...
done
```


</details>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: scripts/generate-proxy-api-models.sh [hunk 9]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+docker run --rm -v "$PWD:/usr/src/app" -e "NODE_ENV=development" -w "/usr/src/app" italia-backend/tools:1.0.0 yarn generate:proxy-api-models
```
Comment (cloudify) [inline scripts/generate-proxy-api-models.sh]: why run this inside a docker container?

Comment (lussoluca) [reply scripts/generate-proxy-api-models.sh]: it is just a shortcut if the user doesn't have node installed locally

**Pull request merged**